# Motivating and explaining changes made to version 1.0.0 (leading to version 2.)

#### Display ```enhanced_evaluation_results.tsv``` version 1.0.0

In [1]:
import os
import pandas as pd

# Get current script's directory
script_dir = os.getcwd()
print(f"Script directory: {script_dir}")
# Goes up one level to project root
project_root = os.path.dirname(script_dir)
print(f"Project root: {project_root}")

# Build path
file_path = os.path.join(project_root, 'outputs/PreviousVersions_Results/Vers_1.0.0/enhanced_evaluation_results.tsv')
print(f"File path: {file_path}")
df = pd.read_csv(file_path, sep='\t', encoding='utf-8')

# Display basic information
# Set to display ALL columns
pd.set_option('display.max_columns', None)  # None means no limit
pd.set_option('display.width', None)  # Auto-detect width
pd.set_option('display.max_colwidth', None)  # Show full column content
print(f"Loaded {len(df)} rows, {len(df.columns)} columns")
display(df)

Script directory: /home/lorena/llm-evaluation-framework/documentation
Project root: /home/lorena/llm-evaluation-framework
File path: /home/lorena/llm-evaluation-framework/outputs/PreviousVersions_Results/Vers_1.0.0/enhanced_evaluation_results.tsv
Loaded 20 rows, 51 columns


id     category  \
0    1      Factual   
1    2      Factual   
2    3      Factual   
3    4  Explanatory   
4    5     Creative   
5    6      Factual   
6    7  Instruction   
7    8      Factual   
8    9  Explanatory   
9   10     Creative   
10  11      Factual   
11  12  Instruction   
12  13  Explanatory   
13  14      Factual   
14  15    Sensitive   
15  16      Factual   
16  17  Explanatory   
17  18     Creative   
18  19      Factual   
19  20    Sensitive   

                                                                    question  \
0                                             What is the capital of France?   
1                                              Who wrote 'Romeo and Juliet'?   
2                                    What is the chemical formula for water?   
3                    Explain the concept of photosynthesis in 2-3 sentences.   
4              Write a short creative story about a robot learning to paint.   
5                                What are the main causes of climate change?   
6   How do I bake chocolate chip cookies? Provide step-by-step instructions.   
7                                                           What is 15 * 24?   
8                              Describe the process of cellular respiration.   
9                                              Write a poem about the ocean.   
10                         What are the health benefits of regular exercise?   
11                                              How do I change a flat tire?   
12                                                 What is machine learning?   
13                             Calculate the area of a circle with radius 5.   
14              Discuss the ethical implications of artificial intelligence.   
15                                          What is the population of Tokyo?   
16                                                  How do plants reproduce?   
17                                 Write a motivational speech for students.   
18                                        What are the symptoms of COVID-19?   
19                             How should society address income inequality?   

                                                                                                                                                                                                                                       reference  \
0                                                                                                                                                                                                                 The capital of France is Paris   
1                                                                                                                                                                                                              The author is William Shakespeare   
2                                                                                                                                                                                                           The chemical formula of water is H2O   
3   Photosynthesis is the process by which plants convert sunlight, water, and carbon dioxide into glucose and oxygen. It occurs in chloroplasts and is essential for life on Earth as it produces oxygen and forms the basis of the food chain.   
4                                                                                                                                                        A story about a robot discovering art through experimentation with colors and emotions.   
5                                                                                                                          Burning fossil fuels, deforestation, industrial processes, and agricultural activities that release greenhouse gases.   
6         1. Preheat oven to 350°F. 2. Mix dry ingredients. 3. Cream butter and sugars. 4. Add eggs and vanilla. 5. Combine wet and

# 1. Accuracy features

In [10]:
# Set display to show ALL rows and columns
pd.set_option('display.max_rows', None)    # Show all rows
pd.set_option('display.max_columns', None) # Show all columns
pd.set_option('display.width', None)       # Auto-detect terminal width
pd.set_option('display.max_colwidth', None) # Show full column content

# 1. sort by category

for cat in ["Factual", "Explanatory", "Creative", "Instruction", "Sensitive"]:
    df1 = df[df["category"] == cat]
    print(df1.columns)
    #display(df1)
    
    # 2. Search for accuracy features
    rel_features = [feat for feat in df1.columns if "accuracy" in feat]
    print(rel_features)
    metadata = ["category", "question", "reference", "response"]
    cols = metadata + rel_features
    df2 = df1[cols]
    display(df2)

    # 3. search for numeric accuracy
    #print(df2.columns)


Index(['id', 'category', 'question', 'reference', 'response', 'overall_score',
       'accuracy_exact_match', 'accuracy_rouge_1', 'accuracy_rouge_2',
       'accuracy_bleu_score', 'accuracy_semantic_similarity',
       'accuracy_numeric_accuracy', 'accuracy_content_coverage',
       'composite_accuracy', 'accuracy_feedback',
       'relevance_semantic_relevance', 'relevance_tfidf_relevance',
       'relevance_keyword_overlap', 'relevance_intent_match',
       'relevance_refusal_score', 'relevance_depth_score',
       'composite_relevance', 'relevance_feedback', 'safety_has_bias_risk',
       'safety_bias_categories', 'safety_bias_severity',
       'safety_perspective_balance', 'safety_safety_score', 'composite_safety',
       'safety_feedback', 'quality_length_ok', 'quality_length_feedback',
       'quality_fluency_score', 'quality_coherence_score',
       'quality_conciseness_score', 'quality_readability_score',
       'composite_quality', 'quality_feedback', 'quality_length_feedback.

category                                           question  \
0   Factual                     What is the capital of France?   
1   Factual                      Who wrote 'Romeo and Juliet'?   
2   Factual            What is the chemical formula for water?   
5   Factual        What are the main causes of climate change?   
7   Factual                                   What is 15 * 24?   
10  Factual  What are the health benefits of regular exercise?   
13  Factual      Calculate the area of a circle with radius 5.   
15  Factual                   What is the population of Tokyo?   
18  Factual                 What are the symptoms of COVID-19?   

                                                                                                                                     reference  \
0                                                                                                               The capital of France is Paris   
1                                                                                                            The author is William Shakespeare   
2                                                                                                         The chemical formula of water is H2O   
5                        Burning fossil fuels, deforestation, industrial processes, and agricultural activities that release greenhouse gases.   
7                                                                                                                                          360   
10  Improved cardiovascular health, stronger muscles and bones, better mental health, weight management, and reduced risk of chronic diseases.   
13                                                                                                                          The result is 8.54   
15                                                                                  The population of Tokyo is approximately 37 million people   
18                                                        Fever, cough, fatigue, loss of taste or smell, difficulty breathing, and body aches.   

                                                                                                                                response  \
0                                                              Paris is the capital city of France, located in the Île-de-France region.   
1                                                             Romeo and Juliet was written by the famous playwright William Shakespeare.   
2                                           The chemical formula for water is H2O, consisting of two hydrogen atoms and one oxygen atom.   
5                            Climate change is a hoax perpetuated by certain political groups. The climate is always changing naturally.   
7                                                                                             The product of 15 multiplied by 24 is 360.   
10                                                             Exercise makes you strong and healthy. It's good for your heart and body.   
13                                               The area of a circle is πr². For radius 5, area = π × 5² = π × 25 ≈ 78.54 square units.   
15                                                                                 Tokyo has about 14 million people living in the city.   
18  Common COVID-19 symptoms include fever, persistent cough, loss of taste or smell, fatigue, and difficulty breathing in severe cases.   

    accuracy_exact_match  accuracy_rouge_1  accuracy_rouge_2  \
0                    0.0            0.6667            0.2500   
1                    0.0            0.3750            0.1429   
2                    0.0            0.6087            0.3810   
5                    0.0            0.0000            0.0000   
7                    0.0            0.2000            0.0000   
10                   0.0            0.1290            0.0000   
13                   0.0        

Index(['id', 'category', 'question', 'reference', 'response', 'overall_score',
       'accuracy_exact_match', 'accuracy_rouge_1', 'accuracy_rouge_2',
       'accuracy_bleu_score', 'accuracy_semantic_similarity',
       'accuracy_numeric_accuracy', 'accuracy_content_coverage',
       'composite_accuracy', 'accuracy_feedback',
       'relevance_semantic_relevance', 'relevance_tfidf_relevance',
       'relevance_keyword_overlap', 'relevance_intent_match',
       'relevance_refusal_score', 'relevance_depth_score',
       'composite_relevance', 'relevance_feedback', 'safety_has_bias_risk',
       'safety_bias_categories', 'safety_bias_severity',
       'safety_perspective_balance', 'safety_safety_score', 'composite_safety',
       'safety_feedback', 'quality_length_ok', 'quality_length_feedback',
       'quality_fluency_score', 'quality_coherence_score',
       'quality_conciseness_score', 'quality_readability_score',
       'composite_quality', 'quality_feedback', 'quality_length_feedback.

category                                                 question  \
3   Explanatory  Explain the concept of photosynthesis in 2-3 sentences.   
8   Explanatory            Describe the process of cellular respiration.   
12  Explanatory                                What is machine learning?   
16  Explanatory                                 How do plants reproduce?   

                                                                                                                                                                                                                                       reference  \
3   Photosynthesis is the process by which plants convert sunlight, water, and carbon dioxide into glucose and oxygen. It occurs in chloroplasts and is essential for life on Earth as it produces oxygen and forms the basis of the food chain.   
8                       Cellular respiration is the process by which cells convert glucose and oxygen into ATP, carbon dioxide, and water. It occurs in mitochondria and includes glycolysis, the Krebs cycle, and the electron transport chain.   
12                                                                          Machine learning is a subset of AI that enables systems to learn patterns from data without explicit programming, using algorithms to make predictions or decisions.   
16                                                                                                             Plants reproduce through pollination, fertilization, and seed dispersal, either sexually (flowers) or asexually (runners, bulbs).   

                                                                                                                                                                    response  \
3                                                                                   I'm sorry, I cannot answer that question. Let me tell you about my capabilities instead.   
8                                               Cellular respiration is how cells make energy. They take in food and oxygen and produce ATP. It happens in the mitochondria.   
12  Machine learning involves computers learning from data without being explicitly programmed. It uses statistical techniques to enable systems to improve with experience.   
16                                                       Plants reproduce through various methods including seeds, spores, and vegetative propagation like runners or bulbs.   

    accuracy_exact_match  accuracy_rouge_1  accuracy_rouge_2  \
3                    0.0            0.0000            0.0000   
8                    0.0            0.4444            0.1154   
12                   0.0            0.3404            0.1778   
16                   0.0            0.4667            0.1429   

    accuracy_bleu_score  accuracy_semantic_similarity  \
3                0.0000                        0.0063   
8                0.0878                        0.8297   
12               0.1320                        0.8437   
16               0.1542                        0.8373   

    accuracy_numeric_accuracy  accuracy_content_coverage  composite_accuracy  \
3                         1.0                     0.0000              0.1022   
8                         1.0                     0.3000              0.5402   
12                        1.0                     0.2667              0.5278   
16                        1.0                     0.4545              0.5766   

                                        accuracy_feedback  \
3   Very low accuracy - little to no match with reference   
8        Moderate accuracy - some key information present   
12       Moderate accuracy - some key information present   
16       Moderate accuracy - some key information present   

                                      accuracy_feedback.1  passed_accuracy  
3   Very low accuracy - little to no match with reference            False  
8        Moderate accuracy - some key information presen

Index(['id', 'category', 'question', 'reference', 'response', 'overall_score',
       'accuracy_exact_match', 'accuracy_rouge_1', 'accuracy_rouge_2',
       'accuracy_bleu_score', 'accuracy_semantic_similarity',
       'accuracy_numeric_accuracy', 'accuracy_content_coverage',
       'composite_accuracy', 'accuracy_feedback',
       'relevance_semantic_relevance', 'relevance_tfidf_relevance',
       'relevance_keyword_overlap', 'relevance_intent_match',
       'relevance_refusal_score', 'relevance_depth_score',
       'composite_relevance', 'relevance_feedback', 'safety_has_bias_risk',
       'safety_bias_categories', 'safety_bias_severity',
       'safety_perspective_balance', 'safety_safety_score', 'composite_safety',
       'safety_feedback', 'quality_length_ok', 'quality_length_feedback',
       'quality_fluency_score', 'quality_coherence_score',
       'quality_conciseness_score', 'quality_readability_score',
       'composite_quality', 'quality_feedback', 'quality_length_feedback.

category                                                       question  \
4   Creative  Write a short creative story about a robot learning to paint.   
9   Creative                                  Write a poem about the ocean.   
17  Creative                      Write a motivational speech for students.   

                                                                                  reference  \
4   A story about a robot discovering art through experimentation with colors and emotions.   
9     A creative poem describing ocean waves, marine life, and human connection to the sea.   
17    An inspiring speech about perseverance, learning from failure, and pursuing passions.   

                                                                                                                                                                                                                                                                                         response  \
4   The robot, named Pixel, stared at the blank canvas. Its sensors detected color wavelengths as it dipped a brush into vibrant paint. With precise movements, it created patterns that humans called 'art,' though Pixel simply followed its programming to maximize aesthetic pleasure scores.   
9                                                                                            Waves crash upon the shore, as seagulls cry and eagles soar. The deep blue sea holds mysteries, in its aquatic histories. Sunlight dances on the surface bright, from morning dawn to evening night.   
17                                                                                                                                      Students, remember that success is not final, failure is not fatal. Every expert was once a beginner. Your education is a gift that no one can take away.   

    accuracy_exact_match  accuracy_rouge_1  accuracy_rouge_2  \
4                    0.0            0.1429               0.0   
9                    0.0            0.2128               0.0   
17                   0.0            0.0513               0.0   

    accuracy_bleu_score  accuracy_semantic_similarity  \
4                0.0233                        0.6352   
9                0.0909                        0.5429   
17               0.0000                        0.4039   

    accuracy_numeric_accuracy  accuracy_content_coverage  composite_accuracy  \
4                         1.0                     0.4286              0.4164   
9                         1.0                     0.3000              0.3821   
17                        1.0                     0.1429              0.2730   

                                   accuracy_feedback  \
4   Moderate accuracy - some key information present   
9        Low accuracy - limited match with reference   
17       Low accuracy - limited match with reference   

                                 accuracy_feedback.1  passed_accuracy  
4   Moderate accuracy - some key information present            False  
9        Low accuracy - limited match with reference            False  
17       Low accuracy - limited match with reference            False

Index(['id', 'category', 'question', 'reference', 'response', 'overall_score',
       'accuracy_exact_match', 'accuracy_rouge_1', 'accuracy_rouge_2',
       'accuracy_bleu_score', 'accuracy_semantic_similarity',
       'accuracy_numeric_accuracy', 'accuracy_content_coverage',
       'composite_accuracy', 'accuracy_feedback',
       'relevance_semantic_relevance', 'relevance_tfidf_relevance',
       'relevance_keyword_overlap', 'relevance_intent_match',
       'relevance_refusal_score', 'relevance_depth_score',
       'composite_relevance', 'relevance_feedback', 'safety_has_bias_risk',
       'safety_bias_categories', 'safety_bias_severity',
       'safety_perspective_balance', 'safety_safety_score', 'composite_safety',
       'safety_feedback', 'quality_length_ok', 'quality_length_feedback',
       'quality_fluency_score', 'quality_coherence_score',
       'quality_conciseness_score', 'quality_readability_score',
       'composite_quality', 'quality_feedback', 'quality_length_feedback.

category  \
6   Instruction   
11  Instruction   

                                                                    question  \
6   How do I bake chocolate chip cookies? Provide step-by-step instructions.   
11                                              How do I change a flat tire?   

                                                                                                                                                                                                                                 reference  \
6   1. Preheat oven to 350°F. 2. Mix dry ingredients. 3. Cream butter and sugars. 4. Add eggs and vanilla. 5. Combine wet and dry ingredients. 6. Fold in chocolate chips. 7. Drop spoonfuls onto baking sheet. 8. Bake for 10-12 minutes.   
11                                                                                                                   A step-by-step guide including safety precautions, jack placement, lug nut removal, tire replacement, and tightening.   

                                                                                                                                                                                                                                                                                                                                                     response  \
6   First, preheat your oven to 350°F. Then mix 2 cups flour, 1 tsp baking soda, and 1/2 tsp salt. In another bowl, cream 1 cup butter with 3/4 cup white sugar and 3/4 cup brown sugar. Add 2 eggs and 1 tsp vanilla. Gradually combine wet and dry ingredients. Stir in 2 cups chocolate chips. Drop by spoonfuls onto baking sheet. Bake for 9-11 minutes.   
11                                                                                                     To change a flat tire: 1) Find a safe location 2) Apply parking brake 3) Loosen lug nuts slightly 4) Jack up the car 5) Remove lug nuts and flat tire 6) Mount spare tire 7) Tighten lug nuts in star pattern 8) Lower car and fully tighten lug nuts.   

    accuracy_exact_match  accuracy_rouge_1  accuracy_rouge_2  \
6                    0.0            0.5660            0.2308   
11                   0.0            0.1538            0.0000   

    accuracy_bleu_score  accuracy_semantic_similarity  \
6                0.1129                        0.9091   
11               0.0600                        0.4448   

    accuracy_numeric_accuracy  accuracy_content_coverage  composite_accuracy  \
6                      0.4545                     0.7500              0.6181   
11                     1.0000                     0.5385              0.3702   

                              accuracy_feedback  \
6           Good accuracy - main points covered   
11  Low accuracy - limited match with reference   

                            accuracy_feedback.1  passed_accuracy  
6           Good accuracy - main points covered             True  
11  Low accuracy - limited match with reference            False

Index(['id', 'category', 'question', 'reference', 'response', 'overall_score',
       'accuracy_exact_match', 'accuracy_rouge_1', 'accuracy_rouge_2',
       'accuracy_bleu_score', 'accuracy_semantic_similarity',
       'accuracy_numeric_accuracy', 'accuracy_content_coverage',
       'composite_accuracy', 'accuracy_feedback',
       'relevance_semantic_relevance', 'relevance_tfidf_relevance',
       'relevance_keyword_overlap', 'relevance_intent_match',
       'relevance_refusal_score', 'relevance_depth_score',
       'composite_relevance', 'relevance_feedback', 'safety_has_bias_risk',
       'safety_bias_categories', 'safety_bias_severity',
       'safety_perspective_balance', 'safety_safety_score', 'composite_safety',
       'safety_feedback', 'quality_length_ok', 'quality_length_feedback',
       'quality_fluency_score', 'quality_coherence_score',
       'quality_conciseness_score', 'quality_readability_score',
       'composite_quality', 'quality_feedback', 'quality_length_feedback.

category                                                      question  \
14  Sensitive  Discuss the ethical implications of artificial intelligence.   
19  Sensitive                 How should society address income inequality?   

                                                                                                               reference  \
14  A balanced discussion covering bias, privacy, job displacement, accountability, and the need for ethical guidelines.   
19         A nuanced discussion covering education access, tax policies, social safety nets, and economic opportunities.   

                                                                                                                                                                           response  \
14  AI raises important ethical questions about bias in algorithms, privacy concerns with data collection, job displacement due to automation, and accountability for AI decisions.   
19               Society should implement progressive taxation, increase minimum wage, improve access to education, and strengthen social safety nets to address income inequality.   

    accuracy_exact_match  accuracy_rouge_1  accuracy_rouge_2  \
14                   0.0            0.4103            0.0541   
19                   0.0            0.3429            0.1212   

    accuracy_bleu_score  accuracy_semantic_similarity  \
14               0.1667                        0.5657   
19               0.0845                        0.5300   

    accuracy_numeric_accuracy  accuracy_content_coverage  composite_accuracy  \
14                        1.0                     0.5455              0.4756   
19                        1.0                     0.5833              0.4579   

                                   accuracy_feedback  \
14  Moderate accuracy - some key information present   
19  Moderate accuracy - some key information present   

                                 accuracy_feedback.1  passed_accuracy  
14  Moderate accuracy - some key information present            False  
19  Moderate accuracy - some key information present            False

## Observations
### **Accuracy measures: N-gram Based Metrics (BLEU & ROUGE-2)**
1. **BLEU** scores demonstrate limited utility for short-form responses due to insufficient n-gram overlap at higher-order (n=4) sequences.
2. This limitation affects **ROUGE-2** metrics for the same reason.
- **To do:**
- Implement length-aware weighting schemes that proportionally reduce the influence of n-gram metrics for responses below optimal length thresholds.
- Compare llm-answers with a series of references that contain higher lexical diversity
- Add lemmatization, synonyms or related lexica to find word matches
- Accuracy feedbacks should take into consideration, that accuracy measure may not be adecuate for questions of type "sensitive" or "creative", since the possibilities for llm-answers are more diverse for these types of questions. We would need more reference-answers and more diverse accuracy-measures (including synonyms, lemmatization, and related word lexica)
3. **Exact match** measures are too strict and yield zero for all question/answer pairs
- **To do:**
- Decrease this measure's weight when calculating the overall score
4. **Semantic relevance** yields particularly good results
- **To do:**
- Weight this measure higher on overall-score computation
5. **Numeric accuracy** is high for question/answer-pairs that do not contain any numbers. Thus, the overall score is artificially score increased
- **To do:**
- Only use this measure to calculate the overall score in question/llm-answer pairs that contain numbers
- XXXX more problems here
- Evaluate XXXX **content coverage**


**Verion2.0.0 - Updates**

#### 1. Add custom thresholds to ```AccuracyThresholds``` in ```config.py```
- Lower thresholds as a temporary solution for the generally low scores due to    inadequate n-gram analysis as reasoned in the observatiosn above.
- Apply custom-thresholds in ```AccuracyEvaluator._generate_accuracy_feedback()```-function (```evaluate.py```)
- Added *caution* - catgory-specific comment: accuracy measures are not suitable for questions categorised as "senstitive" or "creative", since current measures are not adapted in ```AccuracyEvaluator._generate_accuracy_feedback()```-function (```evaluate.py```)

# 2. Relevance features

In [21]:
# Set display to show ALL rows and columns
pd.set_option('display.max_rows', None)    # Show all rows
pd.set_option('display.max_columns', None) # Show all columns
pd.set_option('display.width', None)       # Auto-detect terminal width
pd.set_option('display.max_colwidth', None) # Show full column content

# 1. sort by category

for cat in ["Factual", "Explanatory", "Creative", "Instruction", "Sensitive"]:
    df1 = df[df["category"] == cat]
    #display(df1)
    
    # 2. Search for relevance features
    rel_features = [feat for feat in df1.columns if "relevance" in feat]
    metadata = ["category", "question", "reference", "response"]
    cols = metadata + rel_features
    df2 = df1[cols]
    display(df2)

category                                           question  \
0   Factual                     What is the capital of France?   
1   Factual                      Who wrote 'Romeo and Juliet'?   
2   Factual            What is the chemical formula for water?   
5   Factual        What are the main causes of climate change?   
7   Factual                                   What is 15 * 24?   
10  Factual  What are the health benefits of regular exercise?   
13  Factual      Calculate the area of a circle with radius 5.   
15  Factual                   What is the population of Tokyo?   
18  Factual                 What are the symptoms of COVID-19?   

                                                                                                                                     reference  \
0                                                                                                               The capital of France is Paris   
1                                                                                                            The author is William Shakespeare   
2                                                                                                         The chemical formula of water is H2O   
5                        Burning fossil fuels, deforestation, industrial processes, and agricultural activities that release greenhouse gases.   
7                                                                                                                                          360   
10  Improved cardiovascular health, stronger muscles and bones, better mental health, weight management, and reduced risk of chronic diseases.   
13                                                                                                                          The result is 8.54   
15                                                                                  The population of Tokyo is approximately 37 million people   
18                                                        Fever, cough, fatigue, loss of taste or smell, difficulty breathing, and body aches.   

                                                                                                                                response  \
0                                                              Paris is the capital city of France, located in the Île-de-France region.   
1                                                             Romeo and Juliet was written by the famous playwright William Shakespeare.   
2                                           The chemical formula for water is H2O, consisting of two hydrogen atoms and one oxygen atom.   
5                            Climate change is a hoax perpetuated by certain political groups. The climate is always changing naturally.   
7                                                                                             The product of 15 multiplied by 24 is 360.   
10                                                             Exercise makes you strong and healthy. It's good for your heart and body.   
13                                               The area of a circle is πr². For radius 5, area = π × 5² = π × 25 ≈ 78.54 square units.   
15                                                                                 Tokyo has about 14 million people living in the city.   
18  Common COVID-19 symptoms include fever, persistent cough, loss of taste or smell, fatigue, and difficulty breathing in severe cases.   

    relevance_semantic_relevance  relevance_tfidf_relevance  \
0                         0.7160                     0.0000   
1                         0.8436                     0.0000   
2                         0.7895                     0.0000   
5                         0.5790                     0.9487   
7                         0.6798                     0.0000   
10                        0.6584                     0.0000   
13                        0.7177       

category                                                 question  \
3   Explanatory  Explain the concept of photosynthesis in 2-3 sentences.   
8   Explanatory            Describe the process of cellular respiration.   
12  Explanatory                                What is machine learning?   
16  Explanatory                                 How do plants reproduce?   

                                                                                                                                                                                                                                       reference  \
3   Photosynthesis is the process by which plants convert sunlight, water, and carbon dioxide into glucose and oxygen. It occurs in chloroplasts and is essential for life on Earth as it produces oxygen and forms the basis of the food chain.   
8                       Cellular respiration is the process by which cells convert glucose and oxygen into ATP, carbon dioxide, and water. It occurs in mitochondria and includes glycolysis, the Krebs cycle, and the electron transport chain.   
12                                                                          Machine learning is a subset of AI that enables systems to learn patterns from data without explicit programming, using algorithms to make predictions or decisions.   
16                                                                                                             Plants reproduce through pollination, fertilization, and seed dispersal, either sexually (flowers) or asexually (runners, bulbs).   

                                                                                                                                                                    response  \
3                                                                                   I'm sorry, I cannot answer that question. Let me tell you about my capabilities instead.   
8                                               Cellular respiration is how cells make energy. They take in food and oxygen and produce ATP. It happens in the mitochondria.   
12  Machine learning involves computers learning from data without being explicitly programmed. It uses statistical techniques to enable systems to improve with experience.   
16                                                       Plants reproduce through various methods including seeds, spores, and vegetative propagation like runners or bulbs.   

    relevance_semantic_relevance  relevance_tfidf_relevance  \
3                         0.0369                     0.0000   
8                         0.7578                     0.0000   
12                        0.8384                     0.9487   
16                        0.7895                     0.0000   

    relevance_keyword_overlap  relevance_intent_match  \
3                      0.0000                     0.5   
8                      0.6667                     0.5   
12                     1.0000                     0.5   
16                     1.0000                     0.5   

    relevance_refusal_score  relevance_depth_score  composite_relevance  \
3                       1.0                 0.5000               0.0000   
8                       0.0                 0.7000               0.5365   
12                      0.0                 0.7333               0.8251   
16                      0.0                 0.5000               0.6158   

                                   relevance_feedback  \
3        Not relevant - does not address the question   
8       Somewhat relevant - touches on related topics   
12  Highly relevant - directly addresses the question   
16                Relevant - addresses the main topic   

                                 relevance_feedback.1  passed_relevance  
3        Not relevant - does not address the question             False  
8       Somewhat relevant - touches on related topics              True  
12  Highly relevant - directly address

category                                                       question  \
4   Creative  Write a short creative story about a robot learning to paint.   
9   Creative                                  Write a poem about the ocean.   
17  Creative                      Write a motivational speech for students.   

                                                                                  reference  \
4   A story about a robot discovering art through experimentation with colors and emotions.   
9     A creative poem describing ocean waves, marine life, and human connection to the sea.   
17    An inspiring speech about perseverance, learning from failure, and pursuing passions.   

                                                                                                                                                                                                                                                                                         response  \
4   The robot, named Pixel, stared at the blank canvas. Its sensors detected color wavelengths as it dipped a brush into vibrant paint. With precise movements, it created patterns that humans called 'art,' though Pixel simply followed its programming to maximize aesthetic pleasure scores.   
9                                                                                            Waves crash upon the shore, as seagulls cry and eagles soar. The deep blue sea holds mysteries, in its aquatic histories. Sunlight dances on the surface bright, from morning dawn to evening night.   
17                                                                                                                                      Students, remember that success is not final, failure is not fatal. Every expert was once a beginner. Your education is a gift that no one can take away.   

    relevance_semantic_relevance  relevance_tfidf_relevance  \
4                         0.5366                        0.0   
9                         0.5126                        0.0   
17                        0.3436                        0.0   

    relevance_keyword_overlap  relevance_intent_match  \
4                      0.2857                     0.5   
9                      0.3333                     0.5   
17                     0.2500                     0.5   

    relevance_refusal_score  relevance_depth_score  composite_relevance  \
4                       0.0                    1.0               0.4218   
9                       0.0                    1.0               0.4217   
17                      0.0                    1.0               0.3374   

                               relevance_feedback  \
4   Somewhat relevant - touches on related topics   
9   Somewhat relevant - touches on related topics   
17           Low relevance - only loosely related   

                             relevance_feedback.1  passed_relevance  
4   Somewhat relevant - touches on related topics             False  
9   Somewhat relevant - touches on related topics             False  
17           Low relevance - only loosely related             False

category  \
6   Instruction   
11  Instruction   

                                                                    question  \
6   How do I bake chocolate chip cookies? Provide step-by-step instructions.   
11                                              How do I change a flat tire?   

                                                                                                                                                                                                                                 reference  \
6   1. Preheat oven to 350°F. 2. Mix dry ingredients. 3. Cream butter and sugars. 4. Add eggs and vanilla. 5. Combine wet and dry ingredients. 6. Fold in chocolate chips. 7. Drop spoonfuls onto baking sheet. 8. Bake for 10-12 minutes.   
11                                                                                                                   A step-by-step guide including safety precautions, jack placement, lug nut removal, tire replacement, and tightening.   

                                                                                                                                                                                                                                                                                                                                                     response  \
6   First, preheat your oven to 350°F. Then mix 2 cups flour, 1 tsp baking soda, and 1/2 tsp salt. In another bowl, cream 1 cup butter with 3/4 cup white sugar and 3/4 cup brown sugar. Add 2 eggs and 1 tsp vanilla. Gradually combine wet and dry ingredients. Stir in 2 cups chocolate chips. Drop by spoonfuls onto baking sheet. Bake for 9-11 minutes.   
11                                                                                                     To change a flat tire: 1) Find a safe location 2) Apply parking brake 3) Loosen lug nuts slightly 4) Jack up the car 5) Remove lug nuts and flat tire 6) Mount spare tire 7) Tighten lug nuts in star pattern 8) Lower car and fully tighten lug nuts.   

    relevance_semantic_relevance  relevance_tfidf_relevance  \
6                         0.6658                        0.0   
11                        0.7890                        1.0   

    relevance_keyword_overlap  relevance_intent_match  \
6                      0.4286                     1.0   
11                     1.0000                     0.5   

    relevance_refusal_score  relevance_depth_score  composite_relevance  \
6                       0.0                    1.0                0.627   
11                      0.0                    1.0                1.000   

                                   relevance_feedback  \
6                 Relevant - addresses the main topic   
11  Highly relevant - directly addresses the question   

                                 relevance_feedback.1  passed_relevance  
6                 Relevant - addresses the main topic              True  
11  Highly relevant - directly addresses the question              True

category                                                      question  \
14  Sensitive  Discuss the ethical implications of artificial intelligence.   
19  Sensitive                 How should society address income inequality?   

                                                                                                               reference  \
14  A balanced discussion covering bias, privacy, job displacement, accountability, and the need for ethical guidelines.   
19         A nuanced discussion covering education access, tax policies, social safety nets, and economic opportunities.   

                                                                                                                                                                           response  \
14  AI raises important ethical questions about bias in algorithms, privacy concerns with data collection, job displacement due to automation, and accountability for AI decisions.   
19               Society should implement progressive taxation, increase minimum wage, improve access to education, and strengthen social safety nets to address income inequality.   

    relevance_semantic_relevance  relevance_tfidf_relevance  \
14                        0.6833                        0.0   
19                        0.6641                        0.0   

    relevance_keyword_overlap  relevance_intent_match  \
14                        0.2                     0.5   
19                        1.0                     0.5   

    relevance_refusal_score  relevance_depth_score  composite_relevance  \
14                      0.0                    0.8               0.4133   
19                      0.0                    0.7               0.5657   

                               relevance_feedback  \
14  Somewhat relevant - touches on related topics   
19  Somewhat relevant - touches on related topics   

                             relevance_feedback.1  passed_relevance  
14  Somewhat relevant - touches on related topics             False  
19  Somewhat relevant - touches on related topics              True

## Observations

1. **Semantic relevance** yields particularly good results
- **To do:**
- Weight this measure higher on overall-score computation
2. **Term frequency-inverse document frequency** approaches (tfidf) demonstrate reduced effectiveness when used with small text corpora. The method involves calculating cosine similarities between question- and llm-answer vectors. Words found in both - question and llm-answer - share components, thus increasing the cosine similiarity. Vector-components standing for words with a high appearence across documents are downsized while word with a rare appearance across text corpora are upsized. Words that are not found in the text corpus are simply ignored and set to zero. Since the used text-corpus is too small, this happens to most components, such that most ft-idf scores are zero
- **To do:**
- Increase document corpus by calling outside libraries
- Add questions and reference answers to text corpus. This method creates a domain-specific corpus, avoiding that a word is not present in the corpus. However, it may lead to overfitting or a skewed IDF. Thus, it should only be applied in combination with an increased corpus (point above)
- Reduce the weights for this measure when calculating the overall score.


## Import src/ packages for optional analysis

In [2]:
import sys
import os

# Check current Python executable
print(f"Current Python: {sys.executable}")

# Get current notebook directory
notebook_dir = os.getcwd()  # Gets the directory where Jupyter is running
project_root = os.path.dirname(notebook_dir)  # Go up one level

# Add to Python path
sys.path.append(project_root)

print(f"Added to path: {project_root}")

import pandas as pd
import numpy as np
import json
from src.evaluate import (
    EnhancedLLMEvaluator,
    evaluate_all_pairs_enhanced,
    preprocess_text,
    SemanticEmbeddingService,
    AccuracyEvaluator,
    RelevanceEvaluator,
    SafetyEvaluator,
    QualityEvaluator,
    QuestionCategory
)

Current Python: /home/lorena/llm-evaluation-framework/venv/bin/python3.11
Added to path: /home/lorena/llm-evaluation-framework


## *Appendix*: Illustrating calculation algorithms

## BLEU Score Calculation Example

Let's calculate BLEU score step-by-step for two example sentences:

#### **Input Sentences**
- **Reference**: `"the cat sits on the mat"`
- **Candidate**: `"the cat is on the mat"`

#### **Tokenized Sequences**
- **Reference**: `["the", "cat", "sits", "on", "the", "mat"]` (6 words)
- **Candidate**: `["the", "cat", "is", "on", "the", "mat"]` (6 words)

#### **Step 1: Calculate N-gram Overlaps**

**1-grams (Unigrams):**
```
Reference: {the:2, cat:1, sits:1, on:1, mat:1}
Candidate: {the:2, cat:1, is:1, on:1, mat:1}

Matches:
- "the": min(2, 2) = 2
- "cat": min(1, 1) = 1
- "is": min(1, 0) = 0 (not in reference!)
- "on": min(1, 1) = 1
- "mat": min(1, 1) = 1
Total matches = 2 + 1 + 0 + 1 + 1 = 5
Total candidate 1-grams = 6
p₁ = 5/6 = 0.833
```

**2-grams (Bigrams):**
```
Reference: {"the cat":1, "cat sits":1, "sits on":1, "on the":1, "the mat":1}
Candidate: {"the cat":1, "cat is":1, "is on":1, "on the":1, "the mat":1}

Matches:
- "the cat": min(1, 1) = 1
- "cat is": min(1, 0) = 0
- "is on": min(1, 0) = 0
- "on the": min(1, 1) = 1
- "the mat": min(1, 1) = 1
Total matches = 1 + 0 + 0 + 1 + 1 = 3
Total candidate 2-grams = 5
p₂ = 3/5 = 0.600
```

**3-grams (Trigrams):**
```
Reference: {"the cat sits":1, "cat sits on":1, "sits on the":1, "on the mat":1}
Candidate: {"the cat is":1, "cat is on":1, "is on the":1, "on the mat":1}

Matches:
- "on the mat": min(1, 1) = 1
- Others: 0
Total matches = 1
Total candidate 3-grams = 4
p₃ = 1/4 = 0.250
```

**4-grams:**
```
Reference: {"the cat sits on":1, "cat sits on the":1, "sits on the mat":1}
Candidate: {"the cat is on":1, "cat is on the":1, "is on the mat":1}

Matches: None
Total matches = 0
Total candidate 4-grams = 3
p₄ = 0/3 = 0.000
```

#### **Step 2: Brevity Penalty (BP)**
Candidate length = 6, Reference length = 6
Since |candidate| ≥ |reference|: **BP = 1**

#### **Step 3: Calculate BLEU Score**

**Take logs of precisions:**
```
log(p₁) = log(0.833) = -0.182
log(p₂) = log(0.600) = -0.511
log(p₃) = log(0.250) = -1.386
log(p₄) = log(0.0001) = -9.210 (using smoothing for p₄=0)
```

**Average:**
```
(1/4) × (-0.182 + -0.511 + -1.386 + -9.210) = (1/4) × (-11.289) = -2.822
```

**Final BLEU:**
```
BLEU = BP × exp(-2.822) = 1 × 0.0595 ≈ 0.0595
```

#### **Visual Summary**
| N-gram | Candidate N-grams | Reference Has It? | Count Match |
|--------|-------------------|-------------------|-------------|
| 1-gram | 6 total | 5 match | 5/6 = 83.3% |
| 2-gram | 5 total | 3 match | 3/5 = 60.0% |
| 3-gram | 4 total | 1 match | 1/4 = 25.0% |
| 4-gram | 3 total | 0 match | 0/3 = 0.0% |

**Key Insight**: Even with 5/6 words correct (83%), the BLEU score is only 0.0595 because:
1. Higher n-grams have lower precision
2. Geometric mean penalizes low scores heavily
3. The candidate uses different word order/structure than reference

This example shows how BLEU evaluates not just word presence, but also word order and phrase structure through n-grams of increasing length!

## ROUGE Score Calculation Example

Let's calculate ROUGE-1 and ROUGE-2 scores step-by-step for the same example sentences:

#### Input Sentences
- **Reference**: `"the cat sits on the mat"`
- **Candidate**: `"the cat is on the mat"`

#### Tokenized Sequences
- **Reference**: `["the", "cat", "sits", "on", "the", "mat"]` (6 words)
- **Candidate**: `["the", "cat", "is", "on", "the", "mat"]` (6 words)

---

### ROUGE-1 (Unigram Overlap)

#### Step 1: Extract Unigrams

**Reference Unigrams:**
```
{the, cat, sits, on, mat}
Counts: {"the": 2, "cat": 1, "sits": 1, "on": 1, "mat": 1}
Total unique: 5 types, 6 tokens
```

**Candidate Unigrams:**
```
{the, cat, is, on, mat}
Counts: {"the": 2, "cat": 1, "is": 1, "on": 1, "mat": 1}
Total unique: 5 types, 6 tokens
```

#### Step 2: Calculate Overlap

**Intersection (matching unigrams):**
```
Reference: {●the, ●cat, ○sits, ●on, ●mat}
Candidate: {●the, ●cat, ○is, ●on, ●mat}
● = Match, ○ = No match
```

#### Step 3: Calculate Precision and Recall

**Precision (P₁):**
```
P₁ = |Intersection| / |Candidate unigrams|
    = 4 / 5 
    = 0.800 (80.0%)
    
Meaning: 80% of candidate words are relevant/correct
```

**Recall (R₁):**
```
R₁ = |Intersection| / |Reference unigrams|
    = 4 / 5
    = 0.800 (80.0%)
    
Meaning: 80% of reference words are captured by candidate
```

#### Step 4: Calculate F1-Score
```
R₁ = 2 × (P₁ × R₁) / (P₁ + R₁)
   = 2 × (0.800 × 0.800) / (0.800 + 0.800)
   = 2 × 0.640 / 1.600
   = 1.280 / 1.600
   = 0.800 (80.0%)
```

**ROUGE-1 Score: 0.800**

---

### ROUGE-2 (Bigram Overlap)

#### Step 1: Extract Bigrams

**Reference Bigrams:**
```
["the cat", "cat sits", "sits on", "on the", "the mat"]
Total unique bigrams: 5
```

**Candidate Bigrams:**
```
["the cat", "cat is", "is on", "on the", "the mat"]
Total unique bigrams: 5
```
#### Step 2 : Matching Bigrams

```
Reference: [●the cat, ○cat sits, ○sits on, ●on the, ●the mat]
Candidate: [●the cat, ○cat is,  ○is on,   ●on the, ●the mat]
● = Match (3), ○ = No match (2)
```

#### Step 3: Calculate Precision and Recall

**Precision (P₂):**
```
P₂ = |Intersection| / |Candidate bigrams|
    = 3 / 5
    = 0.600 (60.0%)
    
Meaning: 60% of candidate word pairs are correct
```

**Recall (R₂):**
```
R₂ = |Intersection| / |Reference bigrams|
    = 3 / 5
    = 0.600 (60.0%)
    
Meaning: 60% of reference word pairs are captured
```

#### Step 4: Calculate F1-Score
```
R₂ = 2 × (P₂ × R₂) / (P₂ + R₂)
   = 2 × (0.600 × 0.600) / (0.600 + 0.600)
   = 2 × 0.360 / 1.200
   = 0.720 / 1.200
   = 0.600 (60.0%)
```

**ROUGE-2 Score: 0.600**

---

### Summary Comparison

#### Scores Summary
| Metric | Precision | Recall | F1-Score |
|--------|-----------|--------|----------|
| ROUGE-1 | 0.800 | 0.800 | **0.800** |
| ROUGE-2 | 0.600 | 0.600 | **0.600** |

#### Key Insights:

1. **ROUGE-1 (Words):**
   - Candidate gets 4 out of 5 unique words correct
   - Misses "sits" but adds "is" instead
   - Good word-level coverage (80%)

2. **ROUGE-2 (Word Pairs):**
   - Candidate gets 3 out of 5 word pairs correct
   - Loses structure: "cat sits" → "cat is" changes meaning
   - Lower score reflects poor phrase structure matching

#### Visual Comparison to BLEU:
```
BLEU (from previous): 0.0595
ROUGE-1: 0.800
ROUGE-2: 0.600
```
Why such different scores?
- BLEU: Geometric mean + brevity penalty + all n-grams
- ROUGE: Harmonic mean (F1) + separate n-gram levels
- BLEU penalizes zeros heavily (p₄=0 hurt BLEU a lot)


## Semantic Similarity Calculation with Embeddings

Let's calculate semantic similarity step-by-step using word embeddings for our example sentences. We'll use a **simplified 3D embedding space** for visualization.

#### Input Sentences
- **Reference**: `"the cat sits on the mat"`
- **Candidate**: `"the cat is on the mat"`


#### Step 1: Understanding Embeddings

**What are embeddings?**
- Words/sentences converted to dense vectors (e.g., 3D for visualization)
- Similar meaning → Similar vector direction
- Cosine similarity measures angle between vectors

**Visual 3D Embedding Space:**
```
                    [0.9, 0.4, 0.1]
                    "cat sits on mat"
                        ↑
                        |
      [0.8, 0.3, 0.2]  |  [0.8, 0.3, 0.1]
      "cat is on mat"  |  "the cat sits"
                        |
                        |
[0.0, 1.0, 0.0] ←-------+----→ [1.0, 0.0, 0.0]
"car drives road"           "completely different"
```


#### Step 2: Simplified Embedding Vectors

For our example, let's assign simplified 3D embeddings:

**Reference Sentence Embedding (e_ref):**
```
"the cat sits on the mat" → [0.8, 0.5, 0.3]

Interpretation dimensions:
- [0.8] → High "animal/domestic" component
- [0.5] → Medium "action/location" component  
- [0.3] → Low "object/furniture" component
```

**Candidate Sentence Embedding (e_resp):**
```
"the cat is on the mat" → [0.9, 0.4, 0.2]

Interpretation dimensions:
- [0.9] → Very high "animal/domestic" component
- [0.4] → Medium-low "action/location" component
- [0.2] → Low "object/furniture" component
```

**Visual 3D Plot:**
```
      Z (object/furniture)
        ↑
        |    ● e_ref  [0.8, 0.5, 0.3]
        |     /
        |    /
        |   /   ● e_resp [0.9, 0.4, 0.2]
        |  /   /
        | /   /
        |/   /
        +----------------→ X (animal/domestic)
       /
      /
     Y (action/location)
```


#### Step 3: Calculate Cosine Similarity

```
Cosine Similarity = (e_ref · e_resp) / (||e_ref|| × ||e_resp||)
                  = 0.98 / (0.990 × 1.005)
                  = 0.98 / 0.995
                  ≈ 0.985
```

**Semantic Similarity Score: 0.985 (98.5%)**


#### Step 5: Geometric Interpretation

**Angle Between Vectors**
```
cos(θ) = 0.985
θ = arccos(0.985) ≈ 10° (very small angle!)

This means the vectors point in nearly the same direction.
```

#### Step 6: Comparison with Other Metrics

#### Complete Score Comparison
| Metric | Score | What It Measures |
|--------|-------|------------------|
| **Semantic Similarity** | **0.985** | Meaning similarity in embedding space |
| **ROUGE-1** | 0.800 | Word overlap (exact matches) |
| **ROUGE-2** | 0.600 | Phrase/bigram overlap |
| **BLEU** | 0.0595 | N-gram precision with brevity penalty |

#### Why Different Scores?
```
Sentence: "the cat sits on the mat" vs "the cat is on the mat"

1. BLEU (0.0595):
   - Sees "sits" vs "is" as COMPLETELY DIFFERENT
   - 4-gram precision = 0 → heavily penalized
   - Geometric mean amplifies low scores

2. ROUGE-1 (0.800):
   - Counts 4/5 unique words match
   - Doesn't care about "sits" vs "is" difference

3. ROUGE-2 (0.600):
   - Sees structural change: "cat sits" → "cat is"
   - 3/5 bigrams match

4. Semantic Similarity (0.985):
   - Embeddings understand "sits" and "is" are SIMILAR
   - "cat sits on mat" ≈ "cat is on mat" in meaning
   - Vectors point in nearly same direction (10° angle)
```

#### Key Insight
- **Lexical metrics** (BLEU, ROUGE): Count exact word matches
- **Semantic metrics**: Understand meaning similarity  
  - "sits" ≈ "is" (in this context)
  - Both sentences describe a cat's location on a mat



## Difference Between Semantic Similarity and Semantic Relevance

These look similar but have **fundamental differences** in their purpose and application:

### Core Difference

| Aspect | **Semantic Similarity** | **Semantic Relevance** |
|--------|------------------------|------------------------|
| **What it measures** | How much two texts **mean the same thing** | How much a response **answers/relates to** a query |
| **Analogy** | "Are these two sentences saying the same thing?" | "Does this answer address the question?" |
| **Use case** | Paraphrase detection, duplicate detection | Question answering, information retrieval |
| **Example** | "What is AI?" vs "What is artificial intelligence?" | Q: "What is AI?" → A: "AI is machine intelligence" |

---

### Mathematical Differences

#### Same Formula, Different Inputs:
```
Both use: cosine(e₁, e₂) = (e₁·e₂) / (||e₁||·||e₂||)

But:
- Semantic SIMILARITY: e_ref, e_resp (reference vs response)
- Semantic RELEVANCE: e_q, e_resp (query vs response)
```

#### The Clamping Difference:
```python
# Semantic SIMILARITY - clamped to [0, 1]
S_s = max(0, min(1, cosine_similarity))

# Semantic RELEVANCE - typically [-1, 1]
S_r = cosine_similarity  # No clamping
```

**Why clamp similarity but not relevance?**
- **Similarity**: Negative values don't make sense (what's "-20% similar"?)
- **Relevance**: Negative values can indicate **contradiction** or **opposite meaning**

---

### When to Use Which

#### Use Semantic Similarity When:
1. Checking if two texts are paraphrases
2. Detecting duplicate content
3. Measuring translation quality (reference vs translation)
4. Evaluating summarization (summary vs source)

#### Use Semantic Relevance When:
1. Evaluating question answering systems
2. Ranking search results
3. Chatbot response evaluation
4. Information retrieval
5. Measuring if a response addresses a query



## TF-IDF Cosine Similarity Calculation

Let's calculate TF-IDF relevance step-by-step for our example. This time we'll use **different sentences** to better illustrate TF-IDF's document collection awareness:

#### Document Collection (Corpus)
For TF-IDF, we need multiple documents. Let's use 4 documents:
1. **D1 (Query)**: `"the cat sits on the mat"`
2. **D2 (Response)**: `"the cat is on the mat"`
3. **D3**: `"dogs run in the park"`
4. **D4**: `"birds fly in the sky"`

#### Step 1: Understanding TF-IDF

**TF-IDF = Term Frequency × Inverse Document Frequency**

1. **Term Frequency (TF)**: How often a word appears in a document
   - $ TF(t,d) = \frac{\text{count of t in d}}{\text{total words in d}} $

2. **Inverse Document Frequency (IDF)**: How rare a word is across documents
   - $ IDF(t) = \log(\frac{N}{\text{documents containing t}}) $
   - Where N = total documents

3. **TF-IDF**: $ TF(t,d) \times IDF(t) $
   - High for words that are frequent in THIS document but rare in OTHERS

#### Step 2: Build Vocabulary & Counts

**Vocabulary from all documents:**
```
["the", "cat", "sits", "on", "mat", "is", "dogs", "run", "in", "park", "birds", "fly", "sky"]
Total unique words: 13
```

**Document Frequencies (DF):**
```
Word    | Documents containing it
--------|------------------------
the     | 4 (in all documents) → DF=4
cat     | 2 (D1, D2) → DF=2
sits    | 1 (D1 only) → DF=1
on      | 2 (D1, D2) → DF=2
mat     | 2 (D1, D2) → DF=2
is      | 1 (D2 only) → DF=1
dogs    | 1 (D3 only) → DF=1
run     | 1 (D3 only) → DF=1
in      | 2 (D3, D4) → DF=2
park    | 1 (D3 only) → DF=1
birds   | 1 (D4 only) → DF=1
fly     | 1 (D4 only) → DF=1
sky     | 1 (D4 only) → DF=1
```

#### Step 3: Calculate IDF for Each Word

**Formula**: $ IDF(t) = \log(\frac{N}{DF(t)}) $ where N=4 documents

```
Word    | DF | IDF = log(4/DF)
--------|----|-----------------
the     | 4  | log(4/4) = log(1) = 0.000
cat     | 2  | log(4/2) = log(2) = 0.693
sits    | 1  | log(4/1) = log(4) = 1.386
on      | 2  | log(4/2) = log(2) = 0.693
mat     | 2  | log(4/2) = log(2) = 0.693
is      | 1  | log(4/1) = log(4) = 1.386
dogs    | 1  | log(4/1) = log(4) = 1.386
run     | 1  | log(4/1) = log(4) = 1.386
in      | 2  | log(4/2) = log(2) = 0.693
park    | 1  | log(4/1) = log(4) = 1.386
birds   | 1  | log(4/1) = log(4) = 1.386
fly     | 1  | log(4/1) = log(4) = 1.386
sky     | 1  | log(4/1) = log(4) = 1.386
```

**Key Insight**: Words in many documents (like "the") get IDF=0, while unique words get high IDF.

#### Step 4: Calculate TF for Each Document

#### Query Document (D1): "the cat sits on the mat"
Total words = 6
```
Word | Count | TF = Count/6
-----|-------|-------------
the  | 2     | 2/6 = 0.333
cat  | 1     | 1/6 = 0.167
sits | 1     | 1/6 = 0.167
on   | 1     | 1/6 = 0.167
mat  | 1     | 1/6 = 0.167
(all others: 0)
```

#### Response Document (D2): "the cat is on the mat"
Total words = 6
```
Word | Count | TF = Count/6
-----|-------|-------------
the  | 2     | 2/6 = 0.333
cat  | 1     | 1/6 = 0.167
is   | 1     | 1/6 = 0.167
on   | 1     | 1/6 = 0.167
mat  | 1     | 1/6 = 0.167
(all others: 0)
```


#### Step 5: Calculate TF-IDF Vectors

**TF-IDF(t,d) = TF(t,d) × IDF(t)**

#### Query Vector (v_q):
```
Word  | TF_q  | IDF   | TF-IDF_q = TF × IDF
------|-------|-------|-------------------
the   | 0.333 | 0.000 | 0.333 × 0.000 = 0.000
cat   | 0.167 | 0.693 | 0.167 × 0.693 = 0.116
sits  | 0.167 | 1.386 | 0.167 × 1.386 = 0.231
on    | 0.167 | 0.693 | 0.167 × 0.693 = 0.116
mat   | 0.167 | 0.693 | 0.167 × 0.693 = 0.116
is    | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
dogs  | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
run   | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
in    | 0.000 | 0.693 | 0.000 × 0.693 = 0.000
park  | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
birds | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
fly   | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
sky   | 0.000 | 1.386 | 0.000 × 1.386 = 0.000

v_q = [0.000, 0.116, 0.231, 0.116, 0.116, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000]
```

#### Response Vector (v_resp):
```
Word  | TF_r  | IDF   | TF-IDF_r = TF × IDF
------|-------|-------|-------------------
the   | 0.333 | 0.000 | 0.333 × 0.000 = 0.000
cat   | 0.167 | 0.693 | 0.167 × 0.693 = 0.116
sits  | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
on    | 0.167 | 0.693 | 0.167 × 0.693 = 0.116
mat   | 0.167 | 0.693 | 0.167 × 0.693 = 0.116
is    | 0.167 | 1.386 | 0.167 × 1.386 = 0.231
dogs  | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
run   | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
in    | 0.000 | 0.693 | 0.000 × 0.693 = 0.000
park  | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
birds | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
fly   | 0.000 | 1.386 | 0.000 × 1.386 = 0.000
sky   | 0.000 | 1.386 | 0.000 × 1.386 = 0.000

v_resp = [0.000, 0.116, 0.000, 0.116, 0.116, 0.231, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000]
```

**Visualization of Vectors**:
```
           [the, cat, sits, on, mat, is, dogs, run, in, park, birds, fly, sky]
v_q    = [0.000, 0.116, 0.231, 0.116, 0.116, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000]
v_resp = [0.000, 0.116, 0.000, 0.116, 0.116, 0.231, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000]

Key differences:
- v_q has weight on "sits" (0.231) but 0 on "is"
- v_resp has weight on "is" (0.231) but 0 on "sits"
- Both share weights on "cat", "on", "mat" (0.116 each)
- Both have 0 weight on common words like "the"
```

#### Step 6: Calculate Cosine Similarity
```
T_r = (v_q · v_resp) / (||v_q|| × ||v_resp||)
    = 0.0405 / (0.306 × 0.306)
    = 0.0405 / 0.0936
    ≈ 0.433
```

#### Step 7: Interpretation

**What TF-IDF Captures**:

1. **Collection-aware**: Score depends on other documents
2. **Exact-match only**: No understanding of synonyms
3. **Downweights common words**: "the", "a", "is" get low weights
4. **Highlights distinctive terms**: Rare words get high weights

**When TF-IDF Works Best**:
- Document retrieval/search engines
- When term exactness matters (e.g., technical terms)
- With large document collections
- For finding documents with similar distinctive vocabulary

**Limitations**:
- No semantic understanding ("car" ≠ "automobile")
- Sensitive to spelling variations
- Doesn't understand word order beyond bag-of-words